### ISSUE TITLE
Platform-specific newlines cause spurious diffs with dynamic provider

### ISSUE THREAD

**Issue by hausdorff** [pulumi/pulumi#2240]
In pulumi-kubernetes, a user reports https://github.com/pulumi/pulumi-kubernetes/issues/254:

* If you run `pulumi up` on Windows.
* Then subsequently run `pulumi up` on macos
* Then dynamic provider resources are all updated, because the engine sees textual changes and thinks the field has been updated.

The relevant EKS code is here: https://github.com/pulumi/eks/blob/c6cfed83df8b2b574a21960f7a67cdec6e57ca54/nodejs/eks/transform.ts

Something similar happens in the pulumi-kubernetes YAML classes, and we might decide to normalize newlines to LF in the state, but we'd probably need to provide a path forward for people using with existing YAML state that has CRLF in it. I'm guessing we'll have to make a similar decision for dynamic provider too.

**Comment by hausdorff** [comment]
cc @pgavlin @lukehoban

**Comment by hausdorff** [comment]
Note, as I said in https://github.com/pulumi/pulumi-kubernetes/issues/254, it may not be possible to gracefully fix existing YAML resources from pulumi-kubernetes.

Thankfully, as of yet, it is probably pretty rare -- you'd have to be using Pulumi on two OSs, and you'd have to specifically do `k8s.yaml.ConfigGroup` with options like `{yaml: fs.execSync("your-file.yaml").toString()}` instead of `{files: "your-file.yaml"}`. But, this does mean that if we can't fix it there, this issue will probably lead to us re-opening the discussion around provider-based diffing and how to handle lifecycle of component resources.

**Comment by joeduffy** [comment]
Is this due to closure serialization? /cc @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
We take the code we get from v8 and generate the final file without tweaking the text within it.  This seems like a generalized problem you'd get using git with nearly any tool if you've set git to push/pull files using the current platform's newlines.  I thought general guidance was that if you wanted the same newline behavior across platforms that you set your git settings to be "binary" for files so taht the newlines don't change on OSX/Windows.

If we are generating differences when the user source is different, that seems normal to me.  if we are generating differences when the user source is identical, that seems like something that we could look into.

Note: this is the same policy we have for encodings.  If you happened to save in a non-utf8 encoding for example, and got different results, i think that would be considered by-design.  For identical results, provide identical inputs.

**Comment by CyrusNajmabadi** [comment]
And, specifically, it's not just a general wariness about doing this.  Actually changing the user source is risky.  JS is whitespace sensitive.  You may have, for example, multi-line strings where the newline character is meaningful.  For us to do this, we'd need to be extremely knowledgeable about what was going on.  This is why, for example, the TS compiler does not change your newlines when compiling TS to JS.  It just avoids all those issues and leaves it to the user to provide inputs in a consistent format.

**Comment by joeduffy** [comment]
Excellent points. @hausdorff, does this behavior differ based on whether autocrlf is true or false?

**Comment by joeduffy** [comment]
Also note, I *think* this scenario is a dynamic provider delivered through NPM. Would you expect line endings to change based on platform in that case? (Not sure what NPM's stance here is...)

**Comment by CyrusNajmabadi** [comment]
> I 100% agree with what Cyrus is saying. But I also think that if a user trusts git to handle autocrlf correctly, they might as well trust Pulumi, too.

I'm really really hesitant about that.  It comes down to whether or not that settings means they want pulumi to care about the representation in the repo, or the representation locally. I've always thought of the pulumi as working on my local copy and respecting what is in that.  

Let me put it another way: if they were not using pulumi, but were instead just packing up things into lambdas manually (or using some other tooling), wouldn't tehy also have this problem when bouncing between developers?  The solution, for all those other tools, would likely be that they needed to unify their line endings as well.  So it feels like the most correct, most wide-reaching, approach is to have a consistent set of newline enforcement for those devs, making it so all tools will work consistently.

**Comment by hausdorff** [comment]
> It comes down to whether or not that settings means they want pulumi to care about the representation in the repo, or the representation locally. I've always thought of the pulumi as working on my local copy and respecting what is in that.

The reason git has `autocrlf` is because it makes `diff`, `push`, and `pull` more usable. We have `diff`, `up`, and `new`. Our workflow is a lot closer to git than we are to (say) TypeScript (we both manage an "immutable" state history; theirs is source code, ours is resource definitions) -- I don't understand why we're trying to stick to the latter when users are already trained to use the former.

>  if they were not using pulumi, but were instead just packing up things into lambdas manually (or using some other tooling), wouldn't tehy also have this problem when bouncing between developers?

Sure, but some people also opt out of `autocrlf`. I don't see why this should stop us from adding the option.

> The solution, for all those other tools, would likely be that they needed to unify their line endings as well. So it feels like the most correct, most wide-reaching, approach is to have a consistent set of newline enforcement for those devs, making it so all tools will work consistently.

Why do you have to pick one solution? If your developers prefer a different workflow, turn off `autocrlf`.

**Comment by CyrusNajmabadi** [comment]
@hausdorff Sorry, i think i may have been talking across you and not necessarily understanding your position fully.  Could you clarify the approach you think we should be taking?  For example, i could envision certain types of positions for us to take, including, but not limited to:

1. we detect git's autocrlf option and try to follow that behavior.
2. we supply our own autocrlf option to pulumi itself to allow users to control the behavior of newlines.

Are either of these the approach you were suggesting we take?  If not, could you sum up the approach you do think we should consider?  I have a feeling i'm not on the same page as you, and i'd like to rectify that before responding more to help prevent confusion or cross-talk on my part. :)  Thanks!

**Comment by hausdorff** [comment]
No worries, that's par for the course when you're talking to people. :)

You're exactly right, I'd originally imagined (2). I like the idea of (1) but I have a feeling we'll need to be careful about the UX to avoid confusing people.

**Comment by CyrusNajmabadi** [comment]
If we *were* to do something here, I'd honestly prefer '2' myself as it means users are in total control of the behavior and we avoid the confusion as you mentioned.  I'm still on the fence if we should be doing anything here.  It seems like it would be somewhat strange to me to have this scenario come up and have either of these be the right solution.

To expand on that:  We have a real customer who is saying 'devs are using different machines, and getting diffs when running pulumi that we do not want'.  Why is it a *better* solution to tell them "pass in this --autocrlf option to pulumi" instead of saying "could you just set the newline behavior of git so that all your devs see newlines consistently?".  

in other words, if we think the answer is "the user tells the tool what to do", then i think the best answer would be to use the tool that already understands this problem and has good answers for it.

--

Now: if our answer is "pulumi figures out 'the right thing' (tm), without the user doing anything", then i can see why we'd take on this responsibility.  but if this involves additional steps the user needs to take, then my preference would be that those steps are against the part of the system that is the reason today for why they're ending up with files in the first palce that have differing newlines.

**Comment by CyrusNajmabadi** [comment]
Aside: before we take any additional steps here, can we reach out to the users to at least see if hte git approach is palatable to them or not?  If it is, then there's no issue for us here.  If it's not, then i'd love to know both:

1. why not? 
2. what they'd prefer from pulumi.  i.e. do they want the default behavior of pulumi to be one where we try to normalize (even possibly breaking code), or one where we have our current behavior, but have a switch they can use to tell us how to treat newlines.

**Comment by hausdorff** [comment]
Talking offline, I think @CyrusNajmabadi @pgavlin and I agreed that this is _one_ plausible solution to a problem -- _if the users actually have it_.

So, our proposed strategy now is to wait and see if more users run into this. If they do, we will inquire and see if this actually solves their problems.

**Comment by lukehoban** [comment]
I'm going to close this out, as I believe by default we don't want to do anything here, except to fix the newlines issues in https://github.com/pulumi/pulumi-kubernetes/issues/254.  

We should be super-conservative about being too "smart" about newlines - it will surely backfire.  We should fix clear bugs where we directly cause the problem (like https://github.com/pulumi/pulumi-kubernetes/issues/254).

If we hear repeat feedback on this, we can re-evaluate.
